homeassistant/components/tessie/cover.py

```diff
@@ -0,0 +1,107 @@
+"""Cover platform for Tessie integration."""
+from __future__ import annotations
+
+from typing import Any
+
+from tessie_api import (
+    close_charge_port,
+    close_windows,
+    open_unlock_charge_port,
+    vent_windows,
+)
+
+from homeassistant.components.cover import (
+    CoverDeviceClass,
+    CoverEntity,
+    CoverEntityFeature,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DOMAIN
+from .coordinator import TessieDataUpdateCoordinator
+from .entity import TessieEntity
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the Tessie sensor platform from a config entry."""
+    coordinators = hass.data[DOMAIN][entry.entry_id]
+
+    async_add_entities(
+        Entity(coordinator)
+        for Entity in (
+            TessieWindowEntity,
+            TessieChargePortEntity,
+        )
+        for coordinator in coordinators
+    )
+
+
+class TessieWindowEntity(TessieEntity, CoverEntity):
+    """Cover entity for current charge."""
+
+    _attr_device_class = CoverDeviceClass.WINDOW
+    _attr_supported_features = CoverEntityFeature.OPEN | CoverEntityFeature.CLOSE
+
+    def __init__(self, coordinator: TessieDataUpdateCoordinator) -> None:
+        """Initialize the sensor."""
+        super().__init__(coordinator, "windows")
+
+    @property
+    def is_closed(self) -> bool | None:
+        """Return if the cover is closed or not."""
+        return (
+            self.get("vehicle_state_fd_window") == 0
+            and self.get("vehicle_state_fp_window") == 0
+            and self.get("vehicle_state_rd_window") == 0
+            and self.get("vehicle_state_rp_window") == 0
+        )
+
+    async def async_open_cover(self, **kwargs: Any) -> None:
+        """Open windows."""
+        await self.run(vent_windows)
+        self.set(
+            ("vehicle_state_fd_window", 1),
+            ("vehicle_state_fp_window", 1),
+            ("vehicle_state_rd_window", 1),
+            ("vehicle_state_rp_window", 1),
+        )
+
+    async def async_close_cover(self, **kwargs: Any) -> None:
+        """Close windows."""
+        await self.run(close_windows)
+        self.set(
+            ("vehicle_state_fd_window", 0),
+            ("vehicle_state_fp_window", 0),
+            ("vehicle_state_rd_window", 0),
+            ("vehicle_state_rp_window", 0),
```

Comment:
How come we need to change 4 different states?

Reply:
The car opens and closes all 4 windows with the one function call.

Reply:
But we can't manage each window separately, but we have to still give the separate state of every window for it to work? Odd

Reply:
Yeah I know, we get the state of each window, but cannot control them individually. I could expose binary sensors for each window, but I am not sure how much value that has.